Failure of `Poster` in dry run - withdrawn records

Upon further investigation this does appear to be a separate issue and occurs with the third entry in ~entry_form.tsv~ form_input.tsv. This is DataSpace ID: 115045. From investigation, it seems as though previous DataSpace records were withdrawn for this publication. Specifically Nos. 115045 and 115161. The DataSpace links are here:
1. 115045: https://dataspace.princeton.edu/handle/88435/dsp01g158bm37n
2. 115161: https://dataspace.princeton.edu/handle/88435/dsp01ks65hg284

The record to be used, I believe, should be #115162. See: https://dataspace.princeton.edu/handle/88435/dsp01qj72pb21d

It's unclear to me how often withdrawals occur, but based on my understanding of the workflow, this will always break `Poster`. Specifically, `Scraper` does remove such records from entry_form.tsv. However, since form_input.tsv is manually edited (e.g., to include funding information), this does not occur. I believe a cleaner solution is needed to remove non-existing DataSpace records in form_input.tsv without performing a clean fix as that will remove funding information that are manually entered.

Note: The failed CI is with `Poster` and not `Scraper` module on an `assert`. That is, this does address #44:
```
self.generate_upload_json()
File "Poster.py", line 72, in generate_upload_json
    assert len(dspace_data) == 1, dspace_data
AssertionError: []
Error: Process completed with exit code 1.
```

An empty list was the outcome

_Originally posted by @astrochun in https://github.com/pulibrary/dspace-osti/issues/46#issuecomment-897929226_

--- a/Scraper.py
+++ b/Scraper.py
@@ -6,6 +6,8 @@ import pandas as pd
 
 from os.path import join as pjoin
 
+DSPACE_ID = 'DSpace ID'
+
 
 class Scraper:
     """
@@ -18,6 +20,8 @@ class Scraper:
     :param dspace_scrape: JSON output file containing DataSpace metadata
     :param entry_form_full_path: TSV file containing DataSpace
            records not in OSTI
+    :param form_input_full_path: TSV file containing DataSpace
+           records and DOE metadata for submission
     :param to_upload: JSON output file containing metadata for OSTI upload
     :param redirects: JSON output file containing DOI redirects
 
@@ -30,12 +34,14 @@ class Scraper:
     def __init__(self, data_dir='data', osti_scrape='osti_scrape.json',
                  dspace_scrape='dspace_scrape.json',
                  entry_form_full_path='entry_form.tsv',
+                 form_input_full_path='form_input.tsv',
                  to_upload='dataset_metadata_to_upload.json',
                  redirects='redirects.json'):
 
         self.osti_scrape = pjoin(data_dir, osti_scrape)
         self.dspace_scrape = pjoin(data_dir, dspace_scrape)
         self.entry_form = entry_form_full_path
+        self.form_input = form_input_full_path
         self.to_upload = pjoin(data_dir, to_upload)
         self.redirects = pjoin(data_dir, redirects)
 
@@ -164,7 +170,7 @@ class Scraper:
             to_upload_j = json.load(f)
 
         df = pd.DataFrame()
-        df['DSpace ID'] = [item['id'] for item in to_upload_j]
+        df[DSPACE_ID] = [item['id'] for item in to_upload_j]
         df['Issue Date'] = [[m['value'] for m in item['metadata'] if m['key'] == 'dc.date.issued'][0]
                             for item in to_upload_j]
         df['Title'] = [item['name'] for item in to_upload_j]
@@ -191,12 +197,42 @@ class Scraper:
             print(f"\t{repr(row['Title'])}")
             print(f"\t\t{row['Dataspace Link']}")
 
+    def update_form_input(self):
+        """
+        Update form_input.tsv by adding new records or removing DataSpace
+        records that were removed/withdrawn
+
+        In most cases, this will update form_input.tsv. This further supports CI
+        """
+        if os.path.exists(self.form_input):
+            print(f"File exists. Will update: {self.form_input}")
+
+            entry_df = pd.read_csv(self.entry_form, index_col=DSPACE_ID, sep='\t')
+            input_df = pd.read_csv(self.form_input, index_col=DSPACE_ID, sep='\t')
+            print("Identifying DataSpace records to add and remove ...")
+            entry_id = set(entry_df.index)
+            input_id = set(input_df.index)
+            drops = input_id - entry_id
+            adds = entry_id - input_id
+            commons = entry_id & input_id
+            print(f"Commons records : {len(commons):3}")
+            print(f"New records     : {len(adds):3}")
+            print(f"Records to drop : {len(drops):3}")
+            print(f"Removing : {','.join([str(drop) for drop in drops])} ...")
+            input_df.drop(drops, inplace=True)
+            print(f"Appending : {','.join([str(add) for add in adds])}")
+            revised_df = input_df.append(entry_df.loc[adds])
+            revised_df.to_csv(self.form_input, sep='\t')
+        else:
+            raise FileNotFoundError(f"WARNING: {self.form_input} does not exist!")
+
     def run_pipeline(self, scrape=True):
         if scrape:
             self.get_existing_datasets()
             self.get_dspace_metadata()
         self.get_unposted_metadata()
         self.generate_contract_entry_form()
+        self.update_form_input()
 
 
 if __name__ == '__main__':

--- a/Scraper.py
+++ b/Scraper.py
@@ -222,6 +222,14 @@ class Scraper:
             input_df.drop(drops, inplace=True)
             print(f"Appending : {','.join([str(add) for add in adds])}")
             revised_df = input_df.append(entry_df.loc[adds])
+
+            # Sponsoring organizations is always 'PPPL',
+            # DOE Contact is often the case (for starter), and
+            # Datatype seems to be placeholder - not included in OSTI metadata
+            revised_df.loc[adds, 'Sponsoring Organizations'] = 'PPPL'
+            revised_df.loc[adds, 'DOE Contract'] = 'AC02-09CH11466***'
+            revised_df.loc[adds, 'Datatype'] = 'AS'
+
             revised_df.to_csv(self.form_input, sep='\t')
         else:
             raise FileNotFoundError(f"WARNING: {self.form_input} does not exist!")

--- a/Scraper.py
+++ b/Scraper.py
@@ -226,9 +226,9 @@ class Scraper:
             # Sponsoring organizations is always 'PPPL',
             # DOE Contact is often the case (for starter), and
             # Datatype seems to be placeholder - not included in OSTI metadata
-            revised_df.loc[adds, 'Sponsoring Organizations'] = 'PPPL'
-            revised_df.loc[adds, 'DOE Contract'] = 'AC02-09CH11466***'
-            revised_df.loc[adds, 'Datatype'] = 'AS'
+            revised_df.loc[adds, 'Sponsoring Organizations'] = "USDOE Office of Science (SC)"
+            revised_df.loc[adds, 'DOE Contract'] = "AC02-09CH11466***"
+            revised_df.loc[adds, 'Datatype'] = "AS"
 
             revised_df.to_csv(self.form_input, sep='\t')
         else:

--- a/requirements.txt
+++ b/requirements.txt
@@ -2,4 +2,3 @@ requests
 pandas
 dicttoxml
 git+https://github.com/doecode/ostiapi.git
-    

